### PR TITLE
Fix 5594 - ExceptionHandled + Result is broken

### PR DESCRIPTION
**Pull request by rynowak**
This change ensures that setting ExceptionContext.Result will always
execute if set. The problem with 1.1.0 is that when we had a real short
circuit the wrong set of conditions were checked. I suspect that when you
set ExceptionFilter.Result and didn't short circuit that result filters
were also running (which is a bug).

Added a few tests that verify that the result doesn't trigger result
filters.

I did some general cleanup on this code path to make the state transitions
more clear.

No exception was thrown -> BeginResult
Exception was handled -> ExceptionHandled
Exception was not handled -> gets rethrown

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 1]
```diff
@@ -657,7 +675,7 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
                             return task;
                         }
 
-                        goto case State.ResourceEnd;
+                        goto case State.InvokeEnd;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: This was a bug as well 😲

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 2]
```diff
@@ -672,12 +690,15 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
 
                         if (exceptionContext != null)
                         {
-                            if (exceptionContext.Result != null && !exceptionContext.ExceptionHandled)
+                            if (exceptionContext.Result != null ||
+                                exceptionContext.Exception == null ||
+                                exceptionContext.ExceptionHandled)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: It's possible the table we came up with still has a bug. In MVC 5.x, setting the `Result` alone changes nothing because the `Exception` will still be rethrown. I believe that was also the case in ASP.NET Core MVC 1.0.0. Should I double-check with 1.0.0?

If I'm wrong, this is fine for the patch release. We should revisit the `exceptionContext.Result != null && exceptionContext.Exception != null && !exceptionContext.ExceptionHandled` behaviour for 2.0. It doesn't make sense (and confused our customers) to have the `Result` used when they haven't indicated the `Exception` should be suppressed.

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Follow up: I was incorrect and the ASP.NET Core MVC 1.0.0 behaviour when an `IExceptionFilter` sets `Result` but neither nulls out `Exception` nor sets `ExceptionHandled` is not the same as MVC 5.x: ASP.NET Core 1.0.0 and 1.1.0 both execute the `Result` in this case and neither short-circuits.

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: So, no change needed here. I'd like to discuss suppressing the `Exception` in the "`Result`-only" case for 2.0.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Notice no tests changed to confirm a bug was fixed here. Suggest backing this out as it seems unrelated.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Discussed in person and I will add a test that verifies that we **don't** short circuit when you set result. IE any subsequent exception filters still have a chance to run.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: I can't think of a way that this will manifest in observable bad behavior, but it's part of the code we're fixing and it's just so obviously wrong.

"if the task is completed goto state1; else await the task and then goto state2"

If you still feel strongly then I'm at least going to leave a comment here that this is wrong but we decided not to fix it, and then I'll fix it in dev.

**Comment by rynowak** [comment]
@dougbu updated

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 3]
```diff
@@ -657,6 +675,9 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
                             return task;
                         }
 
+                        // Found this while investigating #5594. This is not right, but we think it's harmless
+                        // so we're leaving it here for the patch release. This should go to InvokeEnd if the
+                        // scope is not Resource because that can only happen when there are no resource filters.
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: 👍 like it.

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: And, of course, my requested change is to double the number of comments like this.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 4]
```diff
@@ -626,11 +632,23 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
...
+                        //
+                        // We need to execute the result (if any) and then exit gracefully which unwinding Resource 
+                        // filters.
+
                         Debug.Assert(state != null);
                         Debug.Assert(_exceptionContext != null);
 
+                        if (_exceptionContext.Result == null)
+                        {
+                            _exceptionContext.Result = new EmptyResult();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: We can remove the Debug.Assert on Line 655 since this ensures Result is never null.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Nah, if I don't get to fix my obvious bug I'm not adding more code cleanup.

**Comment by rynowak** [comment]
Fixed by: 
https://github.com/aspnet/Mvc/commit/531c11df2a1326f257a6d521b328937e8b11db88 (rel/1.1.2)
https://github.com/aspnet/Mvc/commit/af5648c1f7b4668aea77dde8ee1eeb26837e97c0 (dev)
